Use generic external sink

SirMoM opened the pull request:
I did a slight redesign of the modules and the logfile support.
The purpose is to enable people to use custom sinks for the log messages. While still supporting the current Logfile.

Based on this version of the Logger I created a test suite for the logger. Take a look here https://github.com/SirMoM/godot-logger-test

@alketii I'm open for any feedback!

Commit:
Use generic external sink

alketii commented:
FWD: @akien-mga

SirMoM committed:
Apply formatting changes

SirMoM commented:
@akien-mga can you give me any feedback?

akien-mga commented:
I guess that makes sense, but I'm not exactly sure I understand how this is used from custom code. Your test project in https://github.com/SirMoM/godot-logger-test doesn't seem to have any class extending `ExternalSink` (aside from `Logfile`).

akien-mga commented on logger.gd:
```diff
@@ -438,25 +462,25 @@ func get_default_logfile_path():
-	"""Retrieve the dictionary containing all logfiles."""
-	return logfiles
+func get_external_sinks():
+	"""Retrieve the dictionary containing all external sink."""
```
```suggestion
	"""Retrieve the dictionary containing all external sinks."""
```

akien-mga commented on logger.gd:
```diff
@@ -654,12 +678,12 @@ func save_config(configfile = default_configfile_path):
 	config.set_value(PLUGIN_NAME, "max_memory_size", max_memory_size)
 
 	# Logfiles config
```
Should be updated too.

akien-mga commented on logger.gd:
```diff
@@ -703,16 +727,16 @@ func load_config(configfile = default_configfile_path):
 	# Load logfiles config and initialize them
-	logfiles = {}
+	external_sinks = {}
 	for logfile_cfg in config.get_value(PLUGIN_NAME, "logfiles"):
```
You changed the value to `external_sinks` when setting, but not when loading, so isn't this broken?

SirMoM replied:
I guess I did not use persisting the config and I did not notice.
Let me check!

SirMoM replied:
This should be fixed in 028a678

akien-mga commented on logger.gd:
```diff
@@ -438,25 +462,25 @@ func get_default_logfile_path():
 	else:
-		logfiles[logfile_path] = Logfile.new(logfile_path)
-	return logfiles[logfile_path]
+		external_sinks[logfile_path] = Logfile.new(logfile_path, null)
```
Why `null`? Shouldn't it have a default value for the `queue_mode`?

akien-mga commented on logger.gd:
```diff
@@ -224,7 +248,7 @@ const QUEUE_NONE = 0
 const QUEUE_SMART = 2
 
-const FILE_BUFFER_SIZE = 30
+const FILE_BUFFER_SIZE = 30  # * TODO move ? @FILE
```
This should likely be moved to `Logfile` yeah. And the `QUEUE_` constants should likely be moved to `ExternalSink` (likely as an enum since that should work well now in Godot).

SirMoM replied:
This is done in 50ed2bf

akien-mga commented on logger.gd:
```diff
@@ -6,36 +6,60 @@
 
 extends Node  # Needed to work as a singleton
 
+
 ##================##
 ## Inner classes  ##
 ##================##
+class ExternalSink:
```
```suggestion
##================##
## Inner classes  ##
##================##


class ExternalSink:
```

akien-mga commented on logger.gd:
```diff
@@ -6,36 +6,60 @@
+class ExternalSink:
+	var name
+	var queue_mode = QUEUE_NONE
+	var buffer = PoolStringArray()
+	var buffer_idx = 0
```
Is there a reason for this to be in `ExternalSink` instead of `Logfile`? It's not used in the "virtual" API provided by this base class so it's maybe best to leave it up to implementers to define how they want to handle their data to fit the API?

SirMoM replied:
I though every external sink should have a buffer / cache mechanism and so it will be universal usable. But if you think it is not necessary.

SirMoM commented:
> I guess that makes sense, but I'm not exactly sure I understand how this is used from custom code. Your test project in https://github.com/SirMoM/godot-logger-test doesn't seem to have any class extending `ExternalSink` (aside from `Logfile`).

That is right, but I used in https://github.com/SirMoM/logging-console in combination with the logger test. This is so I can have the logs in a different window. But the code isn't quite finished....

Commit:
Fix typo in logger.gd

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

Commit:
Format logger.gd

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

SirMoM committed:
Apply formatting changes

Commit:
Adds config_fields enmum to supress inconsistency errors ^^

I forgot to change a string and this helps to prevent this.
Since all accsess is standardized!

Commit:
Reordering of QUEUE modes and file buffer size.

Commit:
Reformat

SirMoM committed:
Apply formatting changes

akien-mga reviewed (approved):
Looks good to me. I'll squash on merge.

akien-mga commented:
So this "actions-user" that auto-applies formatting changes doesn't seem to trigger a CI build so this PR was merged but actually broke CI.. https://github.com/KOBUGE-Games/godot-logger/runs/2351940297?check_suite_focus=true

I really prefer when style checks are giving developers indications, but developers themselves should amend and test their code to ensure they pass style and lint checks. CI should confirm that the code is good, not fix it for us and then miss half the checks.

SirMoM commented:
Should I remove the auto formatting and just keep the linting?
